### ISSUE TITLE
Unable to nullify string attribute using the update function...

### ISSUE THREAD

**Issue by andrewhoff** [bold-commerce/go-shopify#4]
_From @ckeyes88 on July 17, 2018 16:39_

Currently, string values on any of the data structs have the `omitempty` tag associated with them. Because of this, if I try to set a string value to it's `nil` value `""`, `json.Unmarshal` and `json.Marshal` leave off that attribute. This is a problem when trying to update a string field that I want to unset. 

Steps to reproduce (using a variant as example):
- [ ] Create a variant with `v.InventoryManagement = "shopify"`
- [ ] Set `v.InventoryManagement = ""`
- [ ] Call `shopify.Variant.Updated(v)`

Expected: The returned updated variant should have `v.InventoryManagement == ""` and that variant should no longer be set to have shopify track inventory.

Actual: The returned updated variant has `v.InventoryManagement == "shopify"` and that variant has no changes to it's inventory management in the admin console.

Proposed fix: Convert the string values to string pointers so that they can be explicitly nil-able. 

If I'm thinking about the problem incorrectly, I'd love some feedback on how to unset string properties.

_Copied from original issue: getconversio/go-shopify#104_

**Comment by gcfresh** [comment]
The complication with nullable fields and omit empty is applicable to other fields as well.  For example, a variant update has a `CompareAtPrice` nullable field, which is also `omitempty`.  As a result there currently is no method that allows setting the `CompareAtPrice` to null since null will be omitted.  

This seems to be a very old issue as mentioned [here](https://github.com/google/go-github/issues/19).  

Solutions to the problem are non-trivial and any suggestions are welcome.  Some ideas that have been put forth include:
* Using pointers to nullable types.  This may introduce new dependencies to the library to get all types supported. It would be a breaking change and would require versioning and a large overhaul to the entire library.  
* Define custom types with fields that identify Set, Valid, Value similar to [this](https://www.calhoun.io/how-to-determine-if-a-json-key-has-been-set-to-null-or-not-provided/).
* Creating new endpoint using a new model that would allow patching.  Would still need to determine a suitable storage type for the nullable/omitempty fields.

**Comment by BoldBrandonM** [comment]
Following up to echo the concerns here. This is an issue for us at Bold as well, as it can cause us two types of shortcomings:
- inability to send 'nil' values of non-reference primitive datatypes via clients (cannot set 'false' of a bool, or '0' of an int, etc.)
- inability to clear / remove values via null on the JSON for non-nullable datatypes (cannot set null via string, etc.)

These two cases and combinations of them have caused various issues on our side and will need to be addressed as a part of the next major version due to the breaking nature of the fixes required.

For posterity, the solution we are weighing most towards is a mixture of pointers and nullable types based on the constraints on Shopify (nullable fields will use a nullable type, and fields where nil values are relevant will use a pointer. As much as the concerns about increasing complexity of the library are valid, this seems to be the cleanest and most usable approach out of those listed above without limiting the capabilities of the client.

**Comment by c9845** [comment]
Is the Stripe style method of handling nullable values the solution here?  Stripe uses pointers to values, and then when constructing the struct you use `stripe.String()` or `stripe.Int64()`, both of which return pointers to their underlying values.

For example: https://pkg.go.dev/github.com/stripe/stripe-go#ChargeParams and https://pkg.go.dev/github.com/stripe/stripe-go#String

**Comment by oliver006** [comment]
Yeah, that's a way to handle it. It's a massive change to the library, touching virtually every file and struct and all the tests, something I personally don't have time right now.

**Comment by c9845** [comment]
Oh yeah, its a huge breaking rewrite, I get it.  

The question of implementation also becomes, do you use `*string`, `*int`, etc. for all fields, or only fields that are nullable per the Shopify API?  If this is only implemented for some fields, this becomes even more work since someone would have to have very in-depth knowledge of which fields are nullable or not.  I think just implementing pointers for nearly everything except the definitely-are-not-and-will-never-be nullable fields is the way to go, but regardless this is going to be a pain.  I will try to look into rewriting this, but I would need to carve out a lot of time.

**Comment by oliver006** [comment]
> I think just implementing pointers for nearly everything except the definitely-are-not-and-will-never-be nullable fields is the way to go, but regardless this is going to be a pain.

Agreed on all points. 

Let me know if you're down to take on that larger rewrite and I'll hold off with the 4.0 release but I certainly don't have the time right now so in absence of someone stepping up I would just ship 4.0 without the nullable stuff.

**Comment by oliver006** [comment]
Ok, looks like it's not realistic to change all the fields for the upcoming 4.0 release.

I opened https://github.com/bold-commerce/go-shopify/pull/275 to make at least a few changes. So far this covers fields raised in #182 only.

Any candidates from this thread that should be added? If yes please highlight which field exactly (a lot of field names are repeated in multiple places) and I'll make the change, otherwise we'll table this issue until 5.0 (or just close it).

**Comment by BoldBrandonM** [comment]
I believe there are quite a lot of fields which are affected but given they are not reported explicitly in issues for the project, i'd be fine if we just consider to idiomatically address it across the types in a future v5.0. We might be taking a step backwards in terms of usability if a subset of fields work in a different way.
